File: cuda/scripts/invert.py
import numpy as np

# Define the matrix
matrix = np.array([
    [2.0, 2.0, 3.0, 4.0],
    [2.0, 2.0, 4.0, 5.0],
    [3.0, 4.0, 2.0, 6.0],
    [4.0, 5.0, 6.0, 2.0]
])

# Invert the matrix
try:
    inverse_matrix = np.linalg.inv(matrix)
    print("Inverse of the matrix:")
    print(inverse_matrix)
except np.linalg.LinAlgError:
    print("The matrix is singular and cannot be inverted.")

File: eigen/src/invert.cpp
#include "invert.hpp"
#include <Eigen/Dense>
#include <stdexcept>

// Function to compute the inverse of a matrix in place
void invert_matrix(const Eigen::MatrixXd& A, Eigen::MatrixXd& inverse) {
    // Ensure the matrix is square
    if (A.rows() == 0 || A.rows() != A.cols()) {
        throw std::invalid_argument("Matrix must be square and non-empty.");
    }

    // Compute the inverse using Eigen
    Eigen::FullPivLU<Eigen::MatrixXd> lu(A);

    if (!lu.isInvertible()) {
        throw std::runtime_error("Matrix is singular and cannot be inverted.");
    }

    inverse = lu.inverse();
}
